File: use-cases/merchant-fulfillment/code/java/src/main/java/lambda/ProcessNotificationHandler.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.SQSEvent;
import software.amazon.awssdk.services.sfn.SfnClient;
import software.amazon.awssdk.services.sfn.model.StartExecutionRequest;
import software.amazon.awssdk.services.sfn.model.StartExecutionResponse;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lambda.utils.OrderChangeNotification;
import lambda.utils.NotificationOrderSummary;
import lambda.utils.SPAPINotification;
import lambda.utils.StateMachineInput;

import java.io.IOException;
import java.util.UUID;

import static lambda.utils.Constants.MFN_FULFILLMENT_CODE;
import static lambda.utils.Constants.NOTIFICATION_LEVEL_ORDER_LEVEL;
import static lambda.utils.Constants.NOTIFICATION_PROCESSING_STATUS_ALL;
import static lambda.utils.Constants.NOTIFICATION_TYPE_ORDER_CHANGE;
import static lambda.utils.Constants.REFRESH_TOKEN_ARN_ENV_VARIABLE;
import static lambda.utils.Constants.REGION_CODE_ARN_ENV_VARIABLE;
import static lambda.utils.Constants.STATE_MACHINE_ARN_ENV_VARIABLE;

public class ProcessNotificationHandler implements RequestHandler<SQSEvent, String> {

    public String handleRequest(SQSEvent event, Context context) {
        LambdaLogger logger = context.getLogger();
        logger.log("ProcessNotification Lambda started");

        //Iterate over SQS messages
        //Start a Step Functions workflow execution for unprocessed MFN orders
        for (SQSEvent.SQSMessage message : event.getRecords()) {
            logger.log(String.format("Received new notification: %s", message.getBody()));

            try {
                SPAPINotification notification = mapNotification(message.getBody());

                if (!NOTIFICATION_TYPE_ORDER_CHANGE.equals(notification.getNotificationType())) {
                    logger.log(String.format("Notification type %s skipped", notification.getNotificationType()));
                    continue;
                }

                OrderChangeNotification orderNotification = notification.getPayload().getOrderChangeNotification();
                if (!NOTIFICATION_LEVEL_ORDER_LEVEL.equals(orderNotification.getNotificationLevel())) {
                    logger.log(String.format("Notification level %s skipped", orderNotification.getNotificationLevel()));
                    continue;
                }

                NotificationOrderSummary orderSummary = orderNotification.getSummary();

                //If the order is MFN and has not been processed, start a workflow execution
                if (MFN_FULFILLMENT_CODE.equals(orderSummary.getFulfillmentType())
                        && NOTIFICATION_PROCESSING_STATUS_ALL.contains(orderSummary.getOrderStatus())) {

                    logger.log("Starting state machine execution");
                    String executionArn = startStepFunctionsExecution(orderNotification);
                    logger.log(String.format("State machine successfully started. Execution arn: %s", executionArn));
                } else {
                    logger.log(String.format("Order channel %s and status %s skipped",
                            orderSummary.getFulfillmentType(),
                            orderSummary.getOrderStatus()));
                }
            } catch (IOException e) {
                logger.log(String.format("Message body could not be mapped to a SP-API Notification: %s", e.getMessage()));
            }
        }

        return "Finished processing incoming notifications";
    }

    private SPAPINotification mapNotification(String notificationBody) throws IOException {
        ObjectMapper mapper = new ObjectMapper();
        SPAPINotification notification = mapper.readValue(notificationBody, SPAPINotification.class);

        return notification;
    }

    private String startStepFunctionsExecution(OrderChangeNotification orderNotification) throws JsonProcessingException {
        ObjectMapper mapper = new ObjectMapper();
        StateMachineInput input = getStateMachineInput(orderNotification);
        String inputStr = mapper.writeValueAsString(input);

        StartExecutionRequest request = StartExecutionRequest.builder()
                .stateMachineArn(System.getenv(STATE_MACHINE_ARN_ENV_VARIABLE))
                .name(String.format("%s-%s", orderNotification.getAmazonOrderId(), UUID.randomUUID()))
                .input(inputStr)
                .build();

        SfnClient stepFunctions = SfnClient.builder().build();
        StartExecutionResponse result = stepFunctions.startExecution(request);

        return result.executionArn();
    }

    private StateMachineInput getStateMachineInput(OrderChangeNotification orderNotification) {
        return StateMachineInput.builder()
                .orderId(orderNotification.getAmazonOrderId())
                .refreshToken(System.getenv(REFRESH_TOKEN_ARN_ENV_VARIABLE))
                .regionCode(System.getenv(REGION_CODE_ARN_ENV_VARIABLE))
                .build();
    }
}

File: use-cases/shipping-v2/code/python/src/process_notification_handler.py
import os
import uuid
import json
import boto3
import logging

from src.utils import constants

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def lambda_handler(event, context):
    # Log the input event for the Lambda
    logger.info(f"ProcessNotification Lambda input: {event}")
    
    # Iterate over SQS messages
    # Start a Step Functions workflow execution for unprocessed MFN orders
    for message in event[constants.NOTIFICATION_MESSAGE_KEY_NAME]:
        body_message = message[constants.NOTIFICATION_BODY_KEY_NAME]
        logger.info(f"Received new notification: {body_message}")

        try:
            notification = json.loads(body_message)
            if notification[constants.NOTIFICATION_TYPE_KEY_NAME] != constants.NOTIFICATION_TYPE_ORDER_CHANGE:
                logger.warning(f"Notification type {notification[constants.NOTIFICATION_TYPE_KEY_NAME]} skipped")
                continue

            order_notification = notification[constants.NOTIFICATION_PAYLOAD_KEY_NAME][constants.NOTIFICATION_NAME_KEY_NAME]
            if order_notification[constants.NOTIFICATION_LEVEL_KEY_NAME] != constants.NOTIFICATION_LEVEL_ORDER_LEVEL:
                logger.warning(f"Notification level {order_notification[constants.NOTIFICATION_LEVEL_KEY_NAME]} skipped")
                continue

            order_summary = order_notification[constants.NOTIFICATION_SUMMARY_KEY_NAME]

            # If the order is MFN and has not been processed, start a workflow execution
            if order_summary[constants.NOTIFICATION_FFM_TYPE_KEY_NAME] == constants.SHIPPING_FULFILLMENT_CODE and order_summary[constants.NOTIFICATION_ORDER_STATUS_KEY_NAME] in constants.NOTIFICATION_PROCESSING_STATUS_ALL:
                logger.info("Starting state machine execution")
                execution_arn = start_step_functions_execution(order_notification)
                logger.info(f"State machine successfully started. Execution arn: {execution_arn}")
            else:
                logger.warning(f"Order channel {order_summary[constants.NOTIFICATION_FFM_TYPE_KEY_NAME]} and status {order_summary[constants.NOTIFICATION_ORDER_STATUS_KEY_NAME]} skipped")
        except json.JSONDecodeError as e:
            logger.error(f"Message body could not be mapped to a SP-API Notification: {e}")

    return "Finished processing incoming notifications"


def start_step_functions_execution(order_notification):
    request = {
        'stateMachineArn': os.environ.get(constants.STATE_MACHINE_ARN_ENV_VARIABLE),
        'name': f"{order_notification[constants.NOTIFICATION_AMZ_ORDER_ID_KEY_NAME]}-{str(uuid.uuid4())}",
        'input': json.dumps({
            constants.STATE_MACHINE_KEY_CREDENTIALS: {
                constants.STATE_MACHINE_ORDER_ID_KEY_NAME: order_notification[constants.NOTIFICATION_AMZ_ORDER_ID_KEY_NAME],
                constants.STATE_MACHINE_ONE_CLICK_SHIPMENT_KEY_NAME: os.environ.get(constants.ONE_CLICK_SHIPMENT_ENV_VARIABLE)
            }
        })
    }

    step_functions = boto3.client(constants.AWS_STEP_FUNCTIONS_CLIENT_KEY_NAME)
    result = step_functions.start_execution(**request)

    return result[constants.AWS_STEP_FUNCTIONS_ARN_KEY_NAME]